feat(curriculum): add functional programming lab to new cert (#56800)
freeCodeCamp/freeCodeCamp: curriculum/challenges/english/25-front-end-development/lab-sorting-visualizer/6716249b5405164036fd0b0d.md

---
id: 6716249b5405164036fd0b0d
title: Build a Sorting Visualizer
challengeType: 25
dashedName: build-a-sorting-visualizer
demoType: onClick
---

# --description--

> The Bubble Sort algorithm sorts a sequence of integers by comparing couples of adjacent elements starting from the beginning of the sequence. If the first element is greater then the second one, it swaps them. Then, it proceeds with the following couple. When the sequence ends, it starts a new cycle from the beginning of the sequence, and repeats the process until the elements are sorted.

Reviewer: ```suggestion
The Bubble Sort algorithm sorts a sequence of integers by comparing couples of adjacent elements starting from the beginning of the sequence. If the first element is greater than the second one, it swaps them. Then, it proceeds with the following couple. When the sequence ends, it starts a new cycle from the beginning of the sequence, and repeats the process until the elements are sorted.
```

1. You should have a function named `isOrdered` that takes two integers and returns a boolean indicating if the first integer is less than the second.
1. You should have a function named `swapElements` that takes an array of integers and a numeric index.
1. The `swapElements` function should modify the array in place by swapping the element at the passed index and the following element if the first element is greater than the second.
1. You should have a function named `highligthCurrentEls` that takes an HTML element and a numeric index.
1. The `highligthCurrentEls` function should set the border of the given element's descendants placed at the given index and at the next one to `2px dashed red`.
1. When you click `#generate-btn` you should fill `#starting-array-container` with five `span` elements, each with a random number as its text.
1. After you click `#sort-btn`, `#array-container` should contain a `div` element for each of the steps required by the Bubble Sort algorithm to sort the starting array, including the `div` representing the starting array and a `div` representing the sorted array.

> 1. Each `div` should contain five `span`, each with a number as its text.

Reviewer 2: ```suggestion
1. Each `div` should contain five `span` elements, representing the array in its current state of being sorted.
```

genBtn.click();
const children = document.querySelector("#starting-array").querySelectorAll("span");
assert.lengthOf(children, 5);
Array.from(children).forEach(el => {
    assert.equal(el.tagName, "SPAN");
    const num = Number(el.innerText);
    assert.isAtMost(num, 100);
    assert.isAtLeast(num, 1);
})
```

After you click `#sort-btn`, `#array-container` should contain as many `div` elements as the steps required by the Bubble Sort algorithm to sort the starting array, including the `div` representing the starting array and a `div` representing the sorted array.

```js
const genBtn = document.querySelector("#generate-btn");
const sortBtn = document.querySelector("#sort-btn");
// using randomMocker to be sure that the starting array requires 13 steps to be sorted
const randomMocker = new __helpers.RandomMocker();
randomMocker.mock();
try {    
    genBtn.dispatchEvent(new Event("click"));
    sortBtn.dispatchEvent(new Event("click"));
    const container = document.querySelector("#array-container");
    assert.lengthOf(container.children, 13)
    Array.from(container.children).forEach(el => {assert.equal(el.tagName, "DIV")})
} finally {
    randomMocker.restore();
}
```

> After you click `#sort-btn`, each `div` within `#array-container` should contain five `span`, each with number as its text, and arranged to represent the steps required by Bubble Sort algorithm to sort the starting array.

Reviewer 3: ```suggestion
After you click `#sort-btn`, each `div` within `#array-container` should contain five `span`, each with a number as its text, and arranged to represent the steps required by Bubble Sort algorithm to sort the starting array.
```

> After you click `#sort-btn`, each `div` within `#array-container` should contain five `span`, each with a number as its text, and arranged to represent the steps required by Bubble Sort algorithm to sort the starting array.

Reviewer 4: This test is passing on initial seed code.

```js
const finalArr = [9, 23, 26, 38, 58];
const arrays = [
    [26, 9, 58, 23, 38],
    [9, 26, 58, 23, 38],
    [9, 26, 58, 23, 38],
    [9, 26, 23, 58, 38],
    [9, 26, 23, 38, 58],
    [9, 26, 23, 38, 58],
    finalArr,
    finalArr,
    finalArr,
    finalArr,
    finalArr,
    finalArr,
    finalArr
]
const genBtn = document.querySelector("#generate-btn");
const sortBtn = document.querySelector("#sort-btn");
const randomMocker = new __helpers.RandomMocker();
randomMocker.mock();
try {    
    genBtn.dispatchEvent(new Event("click"));
    sortBtn.dispatchEvent(new Event("click"));

>     const container = document.querySelector("#array-container");

Reviewer 3: ```suggestion
    const container = document.querySelector("#array-container");
    assert.isNotEmpty(container.children);
```
maybe
